tests/integration/conftest.py

```diff
@@ -72,6 +73,35 @@ def _pytest_passed(request):
     ))
 
 
+def _rmtree(path, tries=5):
+    """Remove a directory and its contents.
+
+    Attempt to remove a directory up to ``tries`` times before failing.
+
+    This should be used in-place of ``shutil.rmtree`` which doesn't
+    work reliably on NFS.
+
+    See also:
+        * https://stackoverflow.com/questions/58943374/
+          shutil-rmtree-error-when-trying-to-remove-nfs-mounted-directory
+        * https://bugzilla.redhat.com/show_bug.cgi?id=1362667
+
+    Args:
+        path (pathlib.Path): The directory to remove.
+        tries (int): Maximum number of attempts to make.
+
+    """
+    exc = None
+    for _ in range(tries):
+        try:
+            rmtree(path)
+            return
+        except OSError as exc:
+            # give NFS some thinking time
+            sleep(0.01)
+    raise exc
```

Comment:
I am getting scope problems with this 
![image](https://user-images.githubusercontent.com/37735232/85297523-c0c01b00-b49a-11ea-8fd4-4647e338e6a0.png)
Did you mean to tab that raise into the except?
```suggestion
            raise exc
```

Reply:
Whaaa, but it's defined on line 94!

If you indent the `raise exc` then it will only try once rather than multiple times.

Reply:
Curiously if you do:

```
except OSError as exc_:
    exc = exc_
```

It's fine, I think I've just learned something new about Python3 scoping which was unexpected.

The NFS issue sadly isn't fixed, I think that asyncio might be keeping the file-descriptors alive which will cause NFS to write those pesky `.nfs` files you can't get rid of until the process has died, since the process is pytest and pytest is waiting for NFS its a deadlock.

The solution might coincide with the `os._exit` problem if lucky.

Reply:
I've had a good look at this today. I have tried making use of the rmtree callback function onerror to fix it but so far with no success. There is an interesting explanation here...  https://stackoverflow.com/questions/11228079/python-remove-directory-error-file-exists
So, by that response, if we can close the file in cylc, we shouldn't have the problem.
Looking at the .nfs files created, they look like they're garbled database files...
![image](https://user-images.githubusercontent.com/37735232/85343781-ae1f0380-b4e5-11ea-9daa-26c852e33649.png)
Might help us locate where the offending files are being created and then we can properly close them to avoid the issue. 
Incidentally, when running the integration tests, I'm also getting errors of...
![image](https://user-images.githubusercontent.com/37735232/85343706-744dfd00-b4e5-11ea-8568-3b932aeec60f.png)

Reply:
Ah, I've just re-read your message. 
> since the process is pytest and pytest is waiting for NFS its a deadlock.
I had asssumed the process wasn't pytest as I still can not delete the files manually after pytest has completed its run. This is a pretty frustrating problem!

Reply:
Ok, I have a fix (if temporary until a better one comes along!)...  
`rmtree(path, ignore_errors=True)` 
This will still leave the directories in place if .nfs files are there but at least the tests pass and we can manually delete them whenever we are able?

Reply:
I'm now fairly confident that asyncio is what's keeping the file descriptors open so I'll try and fix that underlying issue which in so-doing will strike off a couple of items under **caveats** in the description.

This needn't delay review but you may want to apply your patch or else just comment out the `_rmtree` bit for testing in the mean time.

Reply:
Whilst we're here, if you see the `n_tries` error, it comes from calling `schd.shutdown` from outside of the `main_loop`. At the moment `shutdown` must be in lock-step with the `main_loop` (else the main loop tries to do things whilst `shutdown` is busy tearing stuff apart).

The NFS/asyncio fix will also fix this 😀.

Reply:
> Whilst we're here, if you see the n_tries error, it comes from calling schd.shutdown from outside of the main_loop. 

Ah good to know :)